### PR TITLE
Adding migration guide for deepdev

### PR DESCRIPTION
**Pull request by luisquintanilla**
This provides guidance on how to migrate to Microsoft.ML.Tokenizers

### CODE REVIEW

**Comment by luisquintanilla** [comment]
@stephentoub @ericstj @tarekgh can you please provide any additional feedback / comments on this. Thanks

File: docs/code/microsoft-ml-tokenizers-migration-guide.md [hunk 1]
```diff
@@ -0,0 +1,19 @@
...
+| --- | --- |
+| [TikTokenizer](https://github.com/microsoft/Tokenizer/blob/2c9ba5d343de52eb27521afef7c0c2f0f76c9c52/Tokenizer_C%23/TokenizerLib/TikTokenizer.cs#L20) | [Tokenizer](https://github.com/dotnet/machinelearning/blob/acced974bea6ed484503a595d87a3e7016c8a558/src/Microsoft.ML.Tokenizers/Tokenizer.cs#L28) |
+| [ITokenizer](https://github.com/microsoft/Tokenizer/blob/2c9ba5d343de52eb27521afef7c0c2f0f76c9c52/Tokenizer_C%23/TokenizerLib/ITokenizer.cs#L7) | [Tokenizer](https://github.com/dotnet/machinelearning/blob/acced974bea6ed484503a595d87a3e7016c8a558/src/Microsoft.ML.Tokenizers/Tokenizer.cs#L28) |
+| [TokenizerBuilder](https://github.com/microsoft/Tokenizer/blob/2c9ba5d343de52eb27521afef7c0c2f0f76c9c52/Tokenizer_C%23/TokenizerLib/TokenizerBuilder.cs#L14) | [TikToken.CreateByModelNameAsync](https://github.com/dotnet/machinelearning/blob/acced974bea6ed484503a595d87a3e7016c8a558/src/Microsoft.ML.Tokenizers/Model/Tiktoken.cs#L778) downloads<br>[Tiktoken.CreateByModelName[Async](Stream)](https://github.com/dotnet/machinelearning/blob/acced974bea6ed484503a595d87a3e7016c8a558/src/Microsoft.ML.Tokenizers/Model/Tiktoken.cs#L106-L180) user provided file stream |
```
Comment (stephentoub) [inline docs/code/microsoft-ml-tokenizers-migration-guide.md:13]: The nested `[` in `[Tiktoken.CreateByModelName[Async]` is breaking the formatting.

File: docs/code/microsoft-ml-tokenizers-migration-guide.md [hunk 2]
```diff
@@ -0,0 +1,19 @@
+# Porting to Microsoft.ML.Tokenizers
+
+This guide provides general guidance on how to migrate from various tokenizer libraries to `Microsoft.ML.Tokenizers` for Tiktoken.
```
Comment (stephentoub) [inline docs/code/microsoft-ml-tokenizers-migration-guide.md:3]: Do we plan to add others beyond deepdev?

Comment (luisquintanilla) [reply docs/code/microsoft-ml-tokenizers-migration-guide.md:3]: Yes. That's why we added the DeepDev guidance as its own section and kept the doc more generic.

File: docs/code/microsoft-ml-tokenizers-migration-guide.md [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+| [ITokenizer](https://github.com/microsoft/Tokenizer/blob/2c9ba5d343de52eb27521afef7c0c2f0f76c9c52/Tokenizer_C%23/TokenizerLib/ITokenizer.cs#L7) | [Tokenizer](https://github.com/dotnet/machinelearning/blob/acced974bea6ed484503a595d87a3e7016c8a558/src/Microsoft.ML.Tokenizers/Tokenizer.cs#L28) |
+| [TokenizerBuilder](https://github.com/microsoft/Tokenizer/blob/2c9ba5d343de52eb27521afef7c0c2f0f76c9c52/Tokenizer_C%23/TokenizerLib/TokenizerBuilder.cs#L14) | [TikToken.CreateByModelNameAsync](https://github.com/dotnet/machinelearning/blob/acced974bea6ed484503a595d87a3e7016c8a558/src/Microsoft.ML.Tokenizers/Model/Tiktoken.cs#L778) downloads<br>[Tiktoken.CreateByModelName[Async](Stream)](https://github.com/dotnet/machinelearning/blob/acced974bea6ed484503a595d87a3e7016c8a558/src/Microsoft.ML.Tokenizers/Model/Tiktoken.cs#L106-L180) user provided file stream |
+
+### General Guidance
+
+- Avoid hard-coding tiktoken regexes and special tokens.  Instead use the appropriate Tiktoken.`CreateByModelNameAsync` method to create the tokenizer from either a downloaded file, or a provided stream.
+- Avoid doing encoding if all you need is token count.  Instead use `Tokenizer.CountTokens`.
```
Comment (tarekgh) [inline docs/code/microsoft-ml-tokenizers-migration-guide.md:18]: >Avoid doing encoding if all you need is token count.  Instead use `Tokenizer.CountTokens`. [](http://example.com/codeflow?start=2&length=90)

Avoid doing encoding if you need the token count or encoded Ids.  Instead use `Tokenizer.CountTokens`. for getting the token count and `Tokenizer.EncodeToIds` for getting the encoded ids.

**Review by tarekgh — APPROVED** [review]
Added minor suggestion. modulo the fixing the broken format, LGTM. Thanks!

File: docs/code/microsoft-ml-tokenizers-migration-guide.md (hunk 3 above)
Comment (luisquintanilla) [reply docs/code/microsoft-ml-tokenizers-migration-guide.md:18]: Fixed

File: docs/code/microsoft-ml-tokenizers-migration-guide.md (hunk 1 above)
Comment (luisquintanilla) [reply docs/code/microsoft-ml-tokenizers-migration-guide.md:13]: Fixed. Thanks.

File: docs/code/microsoft-ml-tokenizers-migration-guide.md [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+| --- | --- |
+| [TikTokenizer](https://github.com/microsoft/Tokenizer/blob/2c9ba5d343de52eb27521afef7c0c2f0f76c9c52/Tokenizer_C%23/TokenizerLib/TikTokenizer.cs#L20) | [Tokenizer](https://github.com/dotnet/machinelearning/blob/acced974bea6ed484503a595d87a3e7016c8a558/src/Microsoft.ML.Tokenizers/Tokenizer.cs#L28) |
+| [ITokenizer](https://github.com/microsoft/Tokenizer/blob/2c9ba5d343de52eb27521afef7c0c2f0f76c9c52/Tokenizer_C%23/TokenizerLib/ITokenizer.cs#L7) | [Tokenizer](https://github.com/dotnet/machinelearning/blob/acced974bea6ed484503a595d87a3e7016c8a558/src/Microsoft.ML.Tokenizers/Tokenizer.cs#L28) |
+| [TokenizerBuilder](https://github.com/microsoft/Tokenizer/blob/2c9ba5d343de52eb27521afef7c0c2f0f76c9c52/Tokenizer_C%23/TokenizerLib/TokenizerBuilder.cs#L14) | [TikToken.CreateByModelNameAsync](https://github.com/dotnet/machinelearning/blob/acced974bea6ed484503a595d87a3e7016c8a558/src/Microsoft.ML.Tokenizers/Model/Tiktoken.cs#L778) downloads<br>[Tiktoken.CreateByModelName(Async/Stream)](https://github.com/dotnet/machinelearning/blob/acced974bea6ed484503a595d87a3e7016c8a558/src/Microsoft.ML.Tokenizers/Model/Tiktoken.cs#L106-L180) user provided file stream |
```
Comment (ericstj) [inline docs/code/microsoft-ml-tokenizers-migration-guide.md:13]: @tarekgh did your recent changes affect this?  Also this one might need an update as well: https://github.com/dotnet/machinelearning/blob/main/src/Microsoft.ML.Tokenizers/PACKAGE.md#how-to-use

Comment (tarekgh) [reply docs/code/microsoft-ml-tokenizers-migration-guide.md:13]: Yes, this now should be:

```
| [TokenizerBuilder](https://github.com/microsoft/Tokenizer/blob/2c9ba5d343de52eb27521afef7c0c2f0f76c9c52/Tokenizer_C%23/TokenizerLib/TokenizerBuilder.cs#L14) | [Tokenizer.CreateTiktokenForModel/Async](https://github.com/dotnet/machinelearning/blob/70e191b3fae444f6625fdc001071de1e2bd1080b/src/Microsoft.ML.Tokenizers/Tokenizer.cs#L298-L330) downloads<br> [Tokenizer.CreateTiktokenForModel(Async/Stream)](https://github.com/dotnet/machinelearning/blob/70e191b3fae444f6625fdc001071de1e2bd1080b/src/Microsoft.ML.Tokenizers/Tokenizer.cs#L222-L296) user provided file stream |
```

Comment (tarekgh) [reply docs/code/microsoft-ml-tokenizers-migration-guide.md:13]: I updated https://github.com/dotnet/machinelearning/blob/main/src/Microsoft.ML.Tokenizers/PACKAGE.md too.

**Review by ericstj — COMMENTED** [review]
LGTM - I merged with Tarek's suggestion.  We can update this more as we go.  Future changes *shouldn't* trigger CI if I got the exclusions right.

**Pull request merged**